lib/charms/prometheus_k8s/v0/prometheus_scrape.py

```diff
@@ -474,6 +476,106 @@ def _sanitize_scrape_configuration(job) -> dict:
     return sanitized_job
 
 
+@dataclasses.dataclass(frozen=True)
+class JujuTopology:
+    """Dataclass for storing and formatting juju topology information."""
+
+    model: str
+    model_uuid: str
+    application: str
+    charm_name: str
+
+    @staticmethod
+    def from_charm(charm):
+        """Factory method for creating the topology dataclass from a given charm."""
+        return JujuTopology(
+            model=charm.model.name,
+            model_uuid=charm.model.uuid,
+            application=charm.model.app.name,
+            charm_name=charm.meta.name,
+        )
+
+    def as_str_short_form(self) -> str:
+        """Format the topology information into a terse string."""
+        return f"{self.model}_{self.model_uuid}_{self.application}"
+
+    def as_str_long_form(self) -> str:
+        """Format the topology information into a verbose string."""
+        return 'juju_model="{}", juju_model_uuid="{}", juju_application="{}"'.format(
+            self.model, self.model_uuid, self.application
+        )
+
+    def as_dict(self) -> dict:
+        """Format the topology information into a dict."""
+        return dataclasses.asdict(self)
+
+    def as_dict_long_form(self):
+        """Format the topology information into a dict with keys having 'juju_' as prefix."""
+        return {
+            "juju_model": self.model,
+            "juju_model_uuid": self.model_uuid,
+            "juju_application": self.application,
+        }
+
+    def render(self, template: str):
+        """Render a juju-topology template string with topology info."""
+        return template.replace("%%juju_topology%%", self.as_str_long_form())
+
+
+def load_alert_rules_from_dir(
+    dir_path: Union[str, Path], topology: JujuTopology, *, recursive: bool = False
+) -> List[dict]:
+    """Load alert rules from rule files.
+
+    All rules from files for the same directory are loaded into a single
+    group. The generated name of this group includes juju topology.
+    By default, only the top directory is scanned; for nested scanning, pass `recursive=True`.
+
+    Args:
+        dir_path: directory containing *.rule files (alert rules without groups).
+        topology: a `JujuTopology` instance.
+        recursive: flag indicating whether to scan for rule files recursively.
+
+    Returns:
+        a list of prometheus alert rule groups.
+    """
+    alerts = defaultdict(list)
+    for path in filter(Path.is_file, Path(dir_path).glob("**/*.rule" if recursive else "*.rule")):
+        relpath = os.path.relpath(os.path.dirname(path), dir_path)
+
+        # Generate group name:
+        #  - prefix, from the relative path of the rule file;
+        #  - name, from juju topology
+        group_name = (
+            f"{'' if relpath == '.' else relpath.replace(os.path.sep, '_') + '_'}"
+            f"{topology.as_str_short_form()}_alerts"
+        )
```

Comment:
This kind of comments sometimes indicates that is worth extracting this lines into a small function, for instance: `_generate_group_name(path, topology)`

Reply:
Not sure about this: relpath makes little sense outside of this function so I am more inclined towards an inner function, wdyt?